### ISSUE TITLE
QUDA on OS X 10.10

### ISSUE THREAD

**Issue by lcosmai** [lattice/quda#226]
Installed CUDA 7.0 on OS X 10.10.3  (MacBook Pro, 1 Nvidia GeForce GT 750 M)

Downloaded QUDA 0.7.0 and configured with:
./configure --prefix=/Users/cosmai/quda --exec-prefix=/Users/cosmai/quda --enable-os=osx --enable-cpu-arch=x86_64 --enable-gpu-arch=sm_30  --disable-wilson-dirac --disable-twisted-mass-dirac --disable-ndeg-twisted-mass-dirac --disable-domain-wall-dirac --enable-hisq-fatlink --enable-gauge-force --enable-hisq-force --with-cuda=/usr/local/cuda CC=/opt/local/bin/gcc-mp-4.9 CXX=/opt/local/bin/gcc-mp-4.9/g++-mp-4.9

QUDA 0.7.0 compilation fails with the following error:
ptxas fatal   : Unresolved extern function '_ZNSt3__13powEdi'
make[1]: *** [unitarize_force_quda.o] Error 255

**Comment by mathiaswagner** [comment]
I have not tried on Mac but is gcc supported by Cuda on Mac OS ? I guess clang is the default cc for OS X.

On Apr 28, 2015, at 09:00, lcosmai <notifications@github.com<mailto:notifications@github.com>> wrote:


Installed CUDA 7.0 on OS X 10.10.3 (MacBook Pro, 1 Nvidia GeForce GT 750 M)

Downloaded QUDA 0.7.0 and configured with:
./configure --prefix=/Users/cosmai/quda --exec-prefix=/Users/cosmai/quda --enable-os=osx --enable-cpu-arch=x86_64 --enable-gpu-arch=sm_30 --disable-wilson-dirac --disable-twisted-mass-dirac --disable-ndeg-twisted-mass-dirac --disable-domain-wall-dirac --enable-hisq-fatlink --enable-gauge-force --enable-hisq-force --with-cuda=/usr/local/cuda CC=/opt/local/bin/gcc-mp-4.9 CXX=/opt/local/bin/gcc-mp-4.9/g++-mp-4.9

QUDA 0.7.0 compilation fails with the following error:
ptxas fatal : Unresolved extern function '_ZNSt3__13powEdi'
make[1]: *** [unitarize_force_quda.o] Error 255

**Comment by azrael417** [comment]
Do you know which routine he is complaining about? Can you find that out? Googling for string fragments such as powEd etc. brings no insight. 



> Am 28.04.2015 um 06:14 schrieb Mathias Wagner <notifications@github.com>:
> 
> I have not tried on Mac but is gcc supported by Cuda on Mac OS ? I guess clang is the default cc for OS X.
>

**Comment by mathiaswagner** [comment]
Ok, I reproduced the issue with CUDA 7.0 on OS X 10.10.3 using clang.
```
/Developer/NVIDIA/CUDA-7.0//bin/nvcc  -O3 -m64 -D__COMPUTE_CAPABILITY__=300 -ftz=true -prec-div=false -prec-sqrt=false -w -DGPU_STAGGERED_DIRAC -DGPU_CLOVER_DIRAC -DGPU_WILSON_DIRAC -DGPU_GAUGE_TOOLS -DGPU_FATLINK  -DGPU_UNITARIZE -DGPU_GAUGE_TOOLS -DGPU_GAUGE_FORCE -DGPU_GAUGE_TOOLS -DGPU_HISQ_FORCE -DGPU_STAGGERED_OPROD -DGPU_GAUGE_TOOLS -DGPU_DIRECT -DBUILD_QDP_INTERFACE -DBUILD_MILC_INTERFACE -arch=sm_30	   -I/Developer/NVIDIA/CUDA-7.0//include -I../include -Idslash_core -I.  unitarize_force_quda.cu -c -o unitarize_force_quda.o
ptxas fatal   : Unresolved extern function '_ZNSt3__13powEdi'
```

I cannot remember having seen this on Linux.

**Comment by mathiaswagner** [comment]
Might be a CUDA bug???

**Comment by nmrcardoso** [comment]
I had a similar problem in a other code with a cuda function in cuda 7.0.
The workaround was to add " -lcudadevrt".

Can you try this, add -lcudadevrt to
LIB = -L$(CUDA_INSTALL_PATH)/lib -lcudart -lcufft -lcurand
in make.in

**Comment by lcosmai** [comment]
In make.inc

     LIB = -L$(CUDA_INSTALL_PATH)/lib -lcudart

modified to:

     LIB = -L$(CUDA_INSTALL_PATH)/lib -lcudart -lcufft -lcurand -lcudadevrt

as suggested.

But I got the same issue:

/usr/local/cuda/bin/nvcc  -O3 -m64 -D__COMPUTE_CAPABILITY__=300 
-ftz=true -prec-div=false -prec-sqrt=false -w -DGPU_STAGGERED_DIRAC 
-DGPU_CLOVER_DIRAC -DGPU_WILSON_DIRAC -DGPU_GAUGE_TOOLS -DGPU_FATLINK  
-DGPU_UNITARIZE -DGPU_GAUGE_TOOLS -DGPU_GAUGE_FORCE -DGPU_GAUGE_TOOLS 
-DGPU_HISQ_FORCE -DGPU_STAGGERED_OPROD -DGPU_GAUGE_TOOLS -DGPU_DIRECT 
-DBUILD_QDP_INTERFACE -DBUILD_MILC_INTERFACE -arch=sm_30       
-I/usr/local/cuda/include -I../include -Idslash_core -I.  
unitarize_force_quda.cu -c -o unitarize_force_quda.o
ptxas fatal   : Unresolved extern function '_ZNSt3__13powEdi'
make[1]: *** [unitarize_force_quda.o] Error 255
make: *** [lib] Error 2



On 29/04/15 04:27, nmrcardoso wrote:
>
> I had a similar problem in a other code with a cuda function in cuda 7.0.
> The workaround was to add " -lcudadevrt".
>
> Can you try this, add -lcudadevrt to
> LIB = -L$(CUDA_INSTALL_PATH)/lib -lcudart -lcufft -lcurand
> in make.in
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/lattice/quda/issues/226#issuecomment-97286007>.
>

-- 
Leonardo Cosmai
INFN Bari
Via Amendola 173
70126 Bari - Italy
office: +39 080 5443207
mobile: +39 340 3580207

**Comment by nmrcardoso** [comment]
Can you try to add in unitarize_force_quda.cu the 
#include <math_functions.hpp>
?

**Comment by nmrcardoso** [comment]
Can you try to add in unitarize_force_quda.cu the 
math_functions.hpp cuda header file?

**Comment by lcosmai** [comment]
Added:

#include <math_functions.h>

in unitarize_force_quda.cu

But even in this case:


/usr/local/cuda/bin/nvcc  -O3 -m64 -D__COMPUTE_CAPABILITY__=300 
-ftz=true -prec-div=false -prec-sqrt=false -w -DGPU_STAGGERED_DIRAC 
-DGPU_CLOVER_DIRAC -DGPU_WILSON_DIRAC -DGPU_GAUGE_TOOLS -DGPU_FATLINK  
-DGPU_UNITARIZE -DGPU_GAUGE_TOOLS -DGPU_GAUGE_FORCE -DGPU_GAUGE_TOOLS 
-DGPU_HISQ_FORCE -DGPU_STAGGERED_OPROD -DGPU_GAUGE_TOOLS -DGPU_DIRECT 
-DBUILD_QDP_INTERFACE -DBUILD_MILC_INTERFACE -arch=sm_30       
-I/usr/local/cuda/include -I../include -Idslash_core -I.  
unitarize_force_quda.cu -c -o unitarize_force_quda.o
ptxas fatal   : Unresolved extern function '_ZNSt3__13powEdi'
make[1]: *** [unitarize_force_quda.o] Error 255
make: *** [lib] Error 2



On 29/04/15 10:22, nmrcardoso wrote:
>
> Can you try to add in unitarize_force_quda.cu the
> math_functions.hpp cuda header file?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/lattice/quda/issues/226#issuecomment-97350595>.
>

-- 
Leonardo Cosmai
INFN Bari
Via Amendola 173
70126 Bari - Italy
office: +39 080 5443207
mobile: +39 340 3580207

**Comment by nmrcardoso** [comment]
try .hpp not .h
since the inline device functions are defined in the .hpp

**Comment by lcosmai** [comment]
Replaced:

#include <math_functions.h>

to:

#include <math_functions.hpp>

But:

/usr/local/cuda/bin/nvcc  -O3 -m64 -D__COMPUTE_CAPABILITY__=300 
-ftz=true -prec-div=false -prec-sqrt=false -w -DGPU_STAGGERED_DIRAC 
-DGPU_CLOVER_DIRAC -DGPU_WILSON_DIRAC -DGPU_GAUGE_TOOLS -DGPU_FATLINK  
-DGPU_UNITARIZE -DGPU_GAUGE_TOOLS -DGPU_GAUGE_FORCE -DGPU_GAUGE_TOOLS 
-DGPU_HISQ_FORCE -DGPU_STAGGERED_OPROD -DGPU_GAUGE_TOOLS -DGPU_DIRECT 
-DBUILD_QDP_INTERFACE -DBUILD_MILC_INTERFACE -arch=sm_30       
-I/usr/local/cuda/include -I../include -Idslash_core -I.  
unitarize_force_quda.cu -c -o unitarize_force_quda.o
ptxas fatal   : Unresolved extern function '_ZNSt3__13powEdi'
make[1]: *** [unitarize_force_quda.o] Error 255
make: *** [lib] Error 2


On 29/04/15 11:13, nmrcardoso wrote:
>
> try .hpp not .h
> since the inline device functions are defined in the .hpp
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/lattice/quda/issues/226#issuecomment-97364838>.
>

-- 
Leonardo Cosmai
INFN Bari
Via Amendola 173
70126 Bari - Italy
office: +39 080 5443207
mobile: +39 340 3580207

**Comment by mathiaswagner** [comment]
I appreciate all your  efforts but after all that work I am inclined to pass that on as a CUDA bug.

It works for Linux. Last thing we can do is give CUDA 6.5 a try (if the installation on my MacBook is still functional I will try that).

**Comment by mathiaswagner** [comment]
```.../CUDA-6.5/bin/nvcc  -O3 -m64 -D__COMPUTE_CAPABILITY__=300 -ftz=true -prec-div=false -prec-sqrt=false -w -DGPU_STAGGERED_DIRAC -DGPU_CLOVER_DIRAC -DGPU_WILSON_DIRAC -DGPU_GAUGE_TOOLS -DGPU_FATLINK  -DGPU_UNITARIZE -DGPU_GAUGE_TOOLS -DGPU_GAUGE_FORCE -DGPU_GAUGE_TOOLS -DGPU_HISQ_FORCE -DGPU_STAGGERED_OPROD -DGPU_GAUGE_TOOLS -DGPU_DIRECT -DBUILD_QDP_INTERFACE -DBUILD_MILC_INTERFACE -arch=sm_30	   -I/Volumes/Macintosh\ HD/Users/mathias-wagner/CUDA-6.5/include -I../include -Idslash_core -I.  unitarize_force_quda.cu -c -o unitarize_force_quda.o```

works for me.

**Comment by mikeaclark** [comment]
So this is a 7.0 bug? If so I'll file it.

**Comment by mathiaswagner** [comment]
Yes, CUDA 7.0 on OS X seems to be affected by this bug.

**Comment by mikeaclark** [comment]
I have created a hotfix to work around this issue.  Pull request #228.

**Comment by mikeaclark** [comment]
A couple of new data points for us to consider:
- This bug affects std::pow(double,int) and std::pow(float::int), but does not affect std::pow(double,double) or std::pow(float,float).
- It has already been fixed internally and will be in the CUDA 7.5 release.
- Given the bug signature, a simple work around that removes any toolkit library dependence is to modify the use of, e.g., std::pow(double,int) to std::(double,double).

For example if we modify the of unitarize_force_quda.cu pow from
```
    template<class Real>
    __device__ __host__
    Real DerivativeCoefficients<Real>::computeC22(const Real & u, const Real & v, const Real & w){
      Real result = - w*pow(u,4)
	- pow(v,2.0)*pow(u,3)
	+ 3*v*w*pow(u,2)
	- 3*pow(w,2)*u;
      return result;
    }
```
to
```
    template<class Real>
    __device__ __host__
    Real DerivativeCoefficients<Real>::computeC22(const Real & u, const Real & v, const Real & w){
      Real result = - w*pow(u,4.0)
	- pow(v,2.0)*pow(u,3.0)
	+ 3*v*w*pow(u,2.0)
	- 3*pow(w,2.0)*u;
      return result;
    }
```
the bug will be circumnavigated in a clean manor.

**Comment by mathiaswagner** [comment]
We have `const Real & u, const Real & v, const Real & w` so instead of putting the implicit double `2.0` we might also add a `static_cast<Real>` to the calls. 
This makes the code harder to read but is slightly more safe - although I believe we always do this step in double.

But I like this fix way more then the preprocessor hack.

**Comment by mikeaclark** [comment]
I'd favour not doing this, primarily for code readability.

This explicit conversion isn't required as per C++ standard.  The compiler will convert the implicit double values to float at compilation stage and not at runtime.  We'd only need it if we went to the level of passing custom types, such as double-double, or some weird fixed-point format.

**Comment by mathiaswagner** [comment]
I thought it would only convert float -> double? I.e. will convert u to double if it's a float?

Anyway, doesn't really matter to me.

**Comment by mikeaclark** [comment]
Ah, I understand what's you're saying now.  

Anyway, I'll update my hotfix branch shortly...

**Comment by mathiaswagner** [comment]
closed with #228
